widget/window_switcher.lua

```diff
@@ -0,0 +1,400 @@
+                                id = 'thumbnail',
+                                resize = true,
+                                horizontal_fit_policy = "fit",
+                                vertical_fit_policy = "fit"
+                            }
+                        },
+                        {
+                            layout  = wibox.layout.fixed.horizontal,
+                            icon_widget(),
+                            {
+                                widget = wibox.container.margin,
+                                left = dpi(6),
+                                right = dpi(14),
+                                -- Add margins to top and bottom in order to force the
+                                -- text to be on a single line, if needed. Might need
+                                -- to adjust them according to font size.
+                                top  = dpi(14),
+                                bottom = dpi(14),
+                                {
+                                    widget = wibox.widget.textbox,
+                                    id     = 'text_role',
+                                    align  = "center",
+                                },
+                            },
+                        }
+                    },
+                }
+            }
+        end
+    end
+
+    s.window_switcher_tasklist = tasklist_widget()
+
+    s.window_switcher_box = awful.popup
+    ({
+        visible = false,
+        ontop = true,
+        screen = s,
+        bg = background,
+        widget =
+        {
+            {
+                s.window_switcher_tasklist,
+                margins = client_margin,
+                widget = wibox.container.margin
+            },
+            bg = beautiful.bg,
+            shape = helpers.shape.rrect(border_radius),
+            widget = wibox.container.background
+        }
+    })
+
+    -- Center window switcher whenever its height changes
+    s.window_switcher_box:connect_signal
+    (
+        "property::width",
+        function()
+            awful.placement.centered(s.window_switcher_box, { honor_workarea = true, honor_padding = true })
+            if s.window_switcher_box.visible and get_num_clients(s) == 0 then
+                window_switcher_hide()
+            end
+        end
+    )
+    s.window_switcher_box:connect_signal
+    (
+        "property::height",
+        function()
+            awful.placement.centered(s.window_switcher_box, { honor_workarea = true, honor_padding = true })
+            if s.window_switcher_box.visible and get_num_clients(s) == 0 then
+                window_switcher_hide()
+            end
+        end
+    )
+end
+
+local enable = function(opts)
+    local opts = opts or {}
+
+    local type = opts.type or "thumbnail"
+    local client_width
+    if type == "thumbnail" then
+        client_width = opts.client_width or dpi(150)
+    else
+        client_width = opts.client_width or dpi(500)
+    end
```

Comment:
Why not like below?
```lua
local client_width = opts.client_width or dpi(type == "thumbnail" and 150 or 500)
```

Reply:
one can argue that the code above yours is readable :smile: